### PR TITLE
refactor: remove dependency on watchman

### PR DESCRIPTION
**Pull request by thsig**
Instead of using FB's watchman library to monitor FS changes in watch
commands, we now use the pure-JS library chokidar.

### CODE REVIEW

**Comment by thsig** [comment]
This PR resolves https://github.com/garden-io/garden/issues/101.

File: src/watch.ts [hunk 1]
```diff
@@ -117,131 +92,144 @@ async function uniqueDependencyModules(module: Module): Promise<Module[]> {
...
-        if (error) {
-          this.ctx.log.error(`Error while executing watcher.command, args were ${args}, error is: ${error}`)
-          rej(error)
-        }
+    const onFileChanged = this.makeFileChangedHandler(modules, changeHandler)
 
-        res(result)
-      })
+    this.watcher = watch(projectRoot, {
+      ignored: /(^|[\/\\])\../, // just ignore dotfiles for now
```
Comment (edvald) [inline src/watch.ts]: Does this implicitly ignore files in `.gitignore`? If not we should probably add that, and could add `.gardenignore` support along the way (which would work the same).

File: src/watch.ts [hunk 2]
```diff
@@ -117,131 +92,144 @@ async function uniqueDependencyModules(module: Module): Promise<Module[]> {
...
+
+  }
+
+  private invalidateCached(module: Module) {
+    // invalidate the cache for anything attached to the module path or upwards in the directory tree
+    const cacheContext = pathToCacheContext(module.path)
+    this.ctx.invalidateCacheUp(cacheContext)
+
+    this.ctx.invalidateCache(cacheContext)
+    this.ctx.invalidateCacheDown(cacheContext)
```
Comment (edvald) [inline src/watch.ts]: Hmm, I don't think you want to call all three of these?

File: src/watch.ts [hunk 3]
```diff
@@ -117,131 +92,144 @@ async function uniqueDependencyModules(module: Module): Promise<Module[]> {
...
+
+      let changedModule: Module | null = null
+
+      for (const module of modules) {
 
-      const changedModuleName = findKey(relModuleRoots,
-        (relPath) => f.name.startsWith(relPath))
+        if (module.path.startsWith(dirPath)) {
+          // at least one module's root dir was removed
+          await this.invalidateCachedForAll()
```
Comment (edvald) [inline src/watch.ts]: Why invalidate for all?

File: src/watch.ts [hunk 4]
```diff
@@ -117,131 +92,144 @@ async function uniqueDependencyModules(module: Module): Promise<Module[]> {
...
-    }
+      if (basename(filePath) === "garden.yml") {
+        await this.invalidateCachedForAll()
+        return changeHandler(null, true)
+      }
+
+      const changedModule = modules.find(m => filePath.startsWith(m.path)) || null
+
+      if (changedModule) {
+        await this.invalidateCachedForAll()
```
Comment (edvald) [inline src/watch.ts]: Why invalidate for all?

File: src/watch.ts (hunk 2 above)
Comment (thsig) [reply src/watch.ts]: Ah, this is a leftover from when I was debugging that I seem to have forgotten to clean up (same for the other two below, https://github.com/garden-io/garden/pull/192#discussion_r199767084 & https://github.com/garden-io/garden/pull/192#discussion_r199767466.

I'll fix these now.

**Pull request merged**